### DISCUSSION TITLE
Need help understanding colors in manifold.

### DISCUSSION THREAD

**Discussion by briansturgill** [elalish/manifold#711] [Q&A] [answered]
1) Am I correct in believing there is not 2d color support?

2) You appear to be putting colors in the second 3 floats in vert_properties. (The first three being vertices).
This appears to be coloring vertices. 
I would have though one would color faces.
What is the color of a face that has 3 different colored vertices?
Clearly I've got something wrong here, I'd appreciate help understanding how it works.

**Comment by elalish** [comment]
Manifold doesn't support colors, it supports arbitrary vertex properties. CrossSection does not yet support properties.

We use vertex properties because they are a lot more general than face properties: vertex colors, vertex normals, UV coordinates, bone weights, etc. It is standard practice in graphics to interpolate all of these properties across triangles, and these properties are defined for most commonly-used triangle formats like glTF. ManifoldCAD.org is designed to make glTF/3MF authoring easy, so it is more opinionated about what properties go where.

**Reply by briansturgill** [reply]
I've been putting better color support in [Piecad](https://github.com/briansturgill/Piecad). (It was only per object.) One thing I've found out is that vertex-based coloring does not
work with 3d Printing.  Bring up any slicer and open a .obj file (with colors) and you'll see just how bad it is. Fortunately, 3mf allows per face coloring.  Now if only 3mf was actually implemented as a standard instead of all the incompatibilities between vendors. I was running through several of the 3mf examples (including one by you), opening them in each of several different slicers.  Every one of them complained about "config" errors.  Several ditched colors in recovering from the "error".  Really, it's just sad. Worse, they don't read each other's files either. Most are using extensions.

**Comment by briansturgill** [comment, marked as answer]
Here is my solution to using Manifold for coloring in Piecad.
I've added some comments to make sense to manifold users.
```
 def color(self, cspec) -> Obj3d:
        """
        Assign the given color to this object.</summary>
        """
       # self.mo in Piecad refers to the underlying manifold3d object.
        mo = self.mo.as_original()
        # So a copy of the original self is made using as_original, such that a new original_id is created and face runs are all to that original_id.
        Obj3d.color_map[mo.original_id()] = _parse_color(cspec)
         # _parse_color returns a 3 tuple that is 8-bit rgb.
        return Obj3d(mo)
```
This next file is what I call internally, but it only depends on lib3mf and manifold3d.
In other words, this file is useful to any Manifold Python user.
The crux of the file is that the color map above is translated into 3mf triangle coloring using manifold3d "runs".
I had to do my own 3mf export because Trimesh's 3mf color support is totally broken.
[_export_3mf.py](https://github.com/user-attachments/files/31201281/_export_3mf.py)

This would also be an answer to [#1479 ]
